lib/db/schema/patch-095-096.sql

```diff
@@ -0,0 +1,304 @@
+-- table, from which we never delete and so which will not be affected by removal.
+-- The second links to `devices` so we need to replace the `ON DELETE CASCADE`
+-- with manual deletion in a couple of stored procedures.
+
+ALTER TABLE deviceCommands
+  DROP FOREIGN KEY deviceCommands_ibfk_1,
+  DROP FOREIGN KEY deviceCommands_ibfk_2,
+ALGORITHM = INPLACE, LOCK = NONE;
+
+
+CREATE PROCEDURE `deleteAccount_16` (
+  IN `uidArg` BINARY(16)
+)
+BEGIN
+  DECLARE EXIT HANDLER FOR SQLEXCEPTION
+  BEGIN
+    ROLLBACK;
+    RESIGNAL;
+  END;
+
+  START TRANSACTION;
+
+  DELETE FROM sessionTokens WHERE uid = uidArg;
+  DELETE FROM keyFetchTokens WHERE uid = uidArg;
+  DELETE FROM accountResetTokens WHERE uid = uidArg;
+  DELETE FROM passwordChangeTokens WHERE uid = uidArg;
+  DELETE FROM passwordForgotTokens WHERE uid = uidArg;
+  DELETE FROM accounts WHERE uid = uidArg;
+  DELETE devices, deviceCommands FROM devices LEFT JOIN deviceCommands
+    ON (deviceCommands.uid = devices.uid AND deviceCommands.deviceId = devices.id)
+    WHERE devices.uid = uidArg;
+  DELETE FROM unverifiedTokens WHERE uid = uidArg;
+  DELETE FROM unblockCodes WHERE uid = uidArg;
+  DELETE FROM emails WHERE uid = uidArg;
+  DELETE FROM signinCodes WHERE uid = uidArg;
+  DELETE FROM totp WHERE uid = uidArg;
+  DELETE FROM recoveryKeys WHERE uid = uidArg;
+  DELETE FROM recoveryCodes WHERE uid = uidArg;
+  DELETE FROM securityEvents WHERE uid = uidArg;
+
+  COMMIT;
+END;
+
+CREATE PROCEDURE `resetAccount_12` (
+  IN `uidArg` BINARY(16),
+  IN `verifyHashArg` BINARY(32),
+  IN `authSaltArg` BINARY(32),
+  IN `wrapWrapKbArg` BINARY(32),
+  IN `verifierSetAtArg` BIGINT UNSIGNED,
+  IN `verifierVersionArg` TINYINT UNSIGNED
+)
+BEGIN
+  DECLARE EXIT HANDLER FOR SQLEXCEPTION
+  BEGIN
+    ROLLBACK;
+    RESIGNAL;
+  END;
+
+  START TRANSACTION;
+
+  DELETE FROM sessionTokens WHERE uid = uidArg;
+  DELETE FROM keyFetchTokens WHERE uid = uidArg;
+  DELETE FROM accountResetTokens WHERE uid = uidArg;
+  DELETE FROM passwordChangeTokens WHERE uid = uidArg;
+  DELETE FROM passwordForgotTokens WHERE uid = uidArg;
+  DELETE devices, deviceCommands FROM devices LEFT JOIN deviceCommands
+    ON (deviceCommands.uid = devices.uid AND deviceCommands.deviceId = devices.id)
+    WHERE devices.uid = uidArg;  DELETE FROM unverifiedTokens WHERE uid = uidArg;
+   UPDATE accounts
+  SET
+    verifyHash = verifyHashArg,
+    authSalt = authSaltArg,
+    wrapWrapKb = wrapWrapKbArg,
+    verifierSetAt = verifierSetAtArg,
+    verifierVersion = verifierVersionArg,
+    profileChangedAt = verifierSetAtArg
+  WHERE uid = uidArg;
+
+  COMMIT;
+END;
+
+-- XXX TODO: when deleting a sessionToken, we delete the corresponding device record.
+-- How should this behave if that device record also has a refreshToken?
+-- Options include:
+--   * leave it alone
+--   * delete the refresh token
+-- For now we keep the current behaviour of deleting the device record.
```

Comment:
@shane-tomlinson here's a data-modelling question for you that come out of this patch - if we have a device with *both* a refreshToken and a sessionToken, and we delete the sessionToken, what should happen to the device record (and to the corresponding refreshToken)?